File: Java/490.java
class Solution {
    int m;
    int n;
    int[] dx = new int[]{0, 0, -1, 1};
    int[] dy = new int[]{1, -1, 0, 0};
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        m = maze.length;
        n = maze[0].length;
        return dfs(start[0], start[1], destination[0], destination[1], maze, new boolean[m][n]);
    }

    private boolean dfs(int i, int j, int di, int dj, int[][] maze, boolean[][] seen){
        if (seen[i][j]){
            return false;
        }
        if (i==di&&j==dj){
            return true;
        }
        seen[i][j]=true;
        for (int k = 0; k < 4; k++){
            int x = i;
            int y = j;
            while(withinBounds(x+dx[k],y+dy[k]) && maze[x+dx[k]][y+dy[k]]==0){
                x+=dx[k];
                y+=dy[k];
            }
            if (dfs(x, y, di, dj, maze, seen)){
                return true;
            }
        }
        return false;
    }

    private boolean withinBounds(int i, int j){
        return i >= 0 && i < m && j >= 0 && j < n;
    }
}

File: C++/490.cpp
class Solution {
public:
    int dx[4]{0, 0, -1, 1};
    int dy[4]{1, -1, 0, 0};
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& dest) {
        int m = size(maze), n = size(maze[0]);
        bool seen[m][n];
        memset(seen, 0, sizeof(seen));
        queue<array<int, 2>> q;
        q.push({start[0], start[1]});
        seen[start[0]][start[1]]=1;

        while(!q.empty()){
            auto top = q.front();
            int x = top[0], y = top[1];
            if (dest[0] == x && dest[1] == y){
                return true;
            }
            for (int i = 0; i < 4; ++i){
                int nx = x, ny = y;
                while(nx >= 0 && ny >= 0 && nx < m && ny < n && maze[nx][ny] != 1){
                    nx += dx[i];
                    ny += dy[i];
                }
                nx -= dx[i];
                ny -= dy[i];
                if (!seen[nx][ny]){
                    seen[nx][ny]=1;
                    q.push({nx, ny});
                }
            }
            q.pop();
        }
        return false;
    }
};
